File: src/main/kotlin/october/october07/P12101.java
package october.october07;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class P12101 {

    static int n, k;
    static int[] history = new int[11];
    static int cnt = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        solve(0, 0);

        if (cnt != k) {
            System.out.println(-1);
        }
    }

    static boolean solve(int node, int step) {

        if (node > n) {
            return false;
        }

        if (node == n) {
            cnt++;
            if (cnt == k) {
                for (int i = 0; i < step - 1; i++) {
                    System.out.print(history[i] + "+");
                }
                System.out.print(history[step - 1]);
                return true;
            }
        }

        for (int i = 1; i <= 3; i++) {
            history[step] = i;
            if (solve(node + i, step + 1)) return true;
        }

        return false;
    }
}

File: src/main/kotlin/october/october07/P12101.kt
package october.october07

private fun main() {
    val (N, K) = readLine()!!.split(" ").map { it.toInt() }
    val res = ArrayList<String>()

    fun go(now: Int, str: String) {
        if (now == N) {
            res.add(str)
            return
        }
        if (now > N) {
            return
        }

        for (i in 1..3) {
            go(now + i, "$str+$i")
        }
    }

    for (i in 1..3) {
        go(i, "$i")
    }

    println(if (res.size < K) -1 else res[K - 1])
}